File: python/tests/apply_pipeline_test.py
#!/usr/bin/env python3
"""
apply_pipeline_test.py

Mirror of connectors/nodejs/test/apply_pipeline.test.js:
  1. Unsupported patch format → fail closed (reason unsupported_patch_format).
  2. Empty extract → no_files_in_fix (never defaults to monitored log).

Run:  python connectors/python/tests/apply_pipeline_test.py
"""

from __future__ import annotations

import asyncio
import os
import sys
import tempfile
import types
import unittest
from pathlib import Path

_CONNECTOR_DIR = Path(__file__).resolve().parent.parent
if str(_CONNECTOR_DIR) not in sys.path:
    sys.path.insert(0, str(_CONNECTOR_DIR))

os.environ["PATCHERLY_AGENT_NOAUTORUN"] = "1"

if "fcntl" not in sys.modules:
    _fcntl_stub = types.ModuleType("fcntl")
    _fcntl_stub.LOCK_EX = 0  # type: ignore[attr-defined]
    _fcntl_stub.LOCK_UN = 0  # type: ignore[attr-defined]
    _fcntl_stub.flock = lambda *a, **k: None  # type: ignore[attr-defined]
    sys.modules["fcntl"] = _fcntl_stub

from patcherly_agent import PatcherlyAgent  # noqa: E402


class ApplyPipelineTest(unittest.IsolatedAsyncioTestCase):
    async def asyncSetUp(self) -> None:
        self.tmp = tempfile.TemporaryDirectory()
        self.root = Path(self.tmp.name).resolve()
        self.backup_root = self.root / "backups"
        self.target_root = self.root / "target"
        self.backup_root.mkdir()
        self.target_root.mkdir()
        os.environ["PATCHERLY_BACKUP_ROOT"] = str(self.backup_root)
        os.environ["PATCHERLY_TARGET_ROOTS"] = str(self.target_root)
        self.agent = PatcherlyAgent(log_file=str(self.root / "agent_logs.txt"))

    async def asyncTearDown(self) -> None:
        self.tmp.cleanup()
        os.environ.pop("PATCHERLY_BACKUP_ROOT", None)
        os.environ.pop("PATCHERLY_TARGET_ROOTS", None)

    async def test_unsupported_patch_format_fail_closed(self) -> None:
        target = self.target_root / "garbage_target.txt"
        target.write_text("unchanged content\n", encoding="utf-8")
        before = target.read_text(encoding="utf-8")
        malformed = "\n".join(
            [
                f"--- a/{target.as_posix()}",
                f"+++ b/{target.as_posix()}",
                "@@@ this is not a real hunk header @@@",
                "~ no actual diff body",
                "",
            ]
        )
        ok, msg, _backup, reason = await self.agent.apply_fix(
            malformed, error_id="test_unsupported_format"
        )
        self.assertFalse(ok)
        self.assertEqual(reason, "unsupported_patch_format")
        self.assertIn("Unsupported patch format", msg)
        self.assertEqual(target.read_text(encoding="utf-8"), before)

    async def test_no_files_in_fix(self) -> None:
        ok, msg, backup, reason = await self.agent.apply_fix(
            "not a diff and no file paths at all", error_id="test_no_files"
        )
        self.assertFalse(ok)
        self.assertEqual(reason, "no_files_in_fix")
        self.assertIsNone(backup)
        self.assertIn("does not reference any files", msg)

    async def test_mid_apply_failure_restores_all_manifest_files(self) -> None:
        file_a = self.target_root / "multi_a.py"
        file_b = self.target_root / "multi_b.py"
        file_a.write_text("a = 1\n", encoding="utf-8")
        file_b.write_text("b = 2\n", encoding="utf-8")
        orig_a = file_a.read_text(encoding="utf-8")
        orig_b = file_b.read_text(encoding="utf-8")
        patch = "\n".join(
            [
                f"--- a/{file_a.as_posix()}",
                f"+++ b/{file_a.as_posix()}",
                "@@ -1,1 +1,1 @@",
                "-a = 1",
                "+a = 99",
                f"--- a/{file_b.as_posix()}",
                f"+++ b/{file_b.as_posix()}",
                "@@ -1,1 +1,1 @@",
                "-b = NO_MATCH",
                "+b = 3",
                "",
            ]
        )
        ok, _msg, backup, _reason = await self.agent.apply_fix(
            patch, error_id="test_mid_apply_multifile"
        )
        self.assertFalse(ok)
        self.assertIsNotNone(backup)
        self.assertEqual(file_a.read_text(encoding="utf-8"), orig_a)
        self.assertEqual(file_b.read_text(encoding="utf-8"), orig_b)


if __name__ == "__main__":
    unittest.main()

File: php/tests/apply_pipeline_test.php
<?php
/**
 * apply_pipeline_test.php
 *
 * Mirror of connectors/nodejs/test/apply_pipeline.test.js for the PHP agent:
 *   1. Unsupported patch format → fail closed (reason unsupported_patch_format).
 *   2. Empty extract → no_files_in_fix (never defaults to monitored log).
 *   3. Source contract: applySimpleFix / simple-fix fallback must be gone.
 *
 * Usage:
 *   php connectors/php/tests/apply_pipeline_test.php
 */

function fail(string $msg): void {
    fwrite(STDERR, "FAIL: {$msg}\n");
    exit(1);
}

$agentPath = realpath(__DIR__ . '/../patcherly_agent.php');
if (!$agentPath || !file_exists($agentPath)) {
    fail('patcherly_agent.php not found');
}
$source = file_get_contents($agentPath);
if ($source === false) {
    fail('could not read patcherly_agent.php');
}

if (strpos($source, 'function applySimpleFix') !== false) {
    fail('applySimpleFix must be deleted (fail-closed contract)');
}
if (strpos($source, 'falling back to simple fix') !== false) {
    fail('simple-fix fallback log string must be gone');
}
if (strpos($source, "reason' => 'unsupported_patch_format'") === false
    && strpos($source, 'unsupported_patch_format') === false) {
    fail('unsupported_patch_format fail-closed reason must be present');
}
if (strpos($source, 'no_files_in_fix') === false) {
    fail('no_files_in_fix reason must be present');
}
// Must not default extract to $this->logFile
if (preg_match('/return\s*!\s*empty\(\s*\$files\s*\)\s*\?\s*\$files\s*:\s*\[\s*\$this->logFile\s*\]/', $source)) {
    fail('extractFilesFromFix must not fall back to $this->logFile');
}

$tmp = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'patcherly-php-apply-' . bin2hex(random_bytes(4));
$backupRoot = $tmp . DIRECTORY_SEPARATOR . 'backups';
$targetRoot = $tmp . DIRECTORY_SEPARATOR . 'target';
if (!mkdir($backupRoot, 0777, true) || !mkdir($targetRoot, 0777, true)) {
    fail('could not create temp dirs');
}

putenv('PATCHERLY_AGENT_NOAUTORUN=1');
putenv('PATCHERLY_BACKUP_ROOT=' . $backupRoot);
putenv('PATCHERLY_TARGET_ROOTS=' . $targetRoot);
putenv('PATCHERLY_IDS_PATH=' . $targetRoot . DIRECTORY_SEPARATOR . 'ids.json');
putenv('PATCHERLY_QUEUE_PATH=' . $targetRoot . DIRECTORY_SEPARATOR . 'queue.jsonl');

$prevCwd = getcwd();
chdir($targetRoot);

require_once $agentPath;

$agent = new PHPAgent();
// Point log file away from repo cwd so we can assert it is untouched.
$logFile = $tmp . DIRECTORY_SEPARATOR . 'agent.log';
file_put_contents($logFile, "log-sentinel\n");
$refLog = new ReflectionProperty($agent, 'logFile');
$refLog->setAccessible(true);
$refLog->setValue($agent, $logFile);

$targetPath = $targetRoot . DIRECTORY_SEPARATOR . 'garbage_target.txt';
file_put_contents($targetPath, "unchanged content\n");
$before = file_get_contents($targetPath);

$malformed = "--- a/" . str_replace('\\', '/', $targetPath) . "\n"
    . "+++ b/" . str_replace('\\', '/', $targetPath) . "\n"
    . "@@@ this is not a real hunk header @@@\n"
    . "~ no actual diff body\n";

$result = $agent->applyFix($malformed, 'test_unsupported_format');
if (($result['success'] ?? true) !== false) {
    fail('must NOT report success on parse failure');
}
if (($result['reason'] ?? '') !== 'unsupported_patch_format') {
    fail('must surface unsupported_patch_format reason, got ' . var_export($result['reason'] ?? null, true));
}
$after = file_get_contents($targetPath);
if ($after !== $before) {
    fail('target file must remain untouched on parse failure');
}

$empty = $agent->applyFix('not a diff and no file paths at all', 'test_no_files');
if (($empty['success'] ?? true) !== false) {
    fail('empty extract must fail');
}
if (($empty['reason'] ?? '') !== 'no_files_in_fix') {
    fail('empty extract must use no_files_in_fix');
}
if (!empty($empty['backup_metadata'])) {
    fail('empty extract must not create backup metadata');
}
if (file_get_contents($logFile) !== "log-sentinel\n") {
    fail('monitored log must not be overwritten on empty extract');
}

// Mid-apply multifile: first file applies, second hunk mismatches → restore both.
$fileA = $targetRoot . DIRECTORY_SEPARATOR . 'multi_a.php';
$fileB = $targetRoot . DIRECTORY_SEPARATOR . 'multi_b.php';
file_put_contents($fileA, "<?php\n\$a = 1;\n");
file_put_contents($fileB, "<?php\n\$b = 2;\n");
$origA = file_get_contents($fileA);
$origB = file_get_contents($fileB);
$aPosix = str_replace('\\', '/', $fileA);
$bPosix = str_replace('\\', '/', $fileB);
$multiPatch = "--- a/{$aPosix}\n"
    . "+++ b/{$aPosix}\n"
    . "@@ -1,2 +1,2 @@\n"
    . " <?php\n"
    . "-\$a = 1;\n"
    . "+\$a = 99;\n"
    . "--- a/{$bPosix}\n"
    . "+++ b/{$bPosix}\n"
    . "@@ -1,2 +1,2 @@\n"
    . " <?php\n"
    . "-\$b = NO_MATCH;\n"
    . "+\$b = 3;\n";
$mid = $agent->applyFix($multiPatch, 'test_mid_apply_multifile');
if (($mid['success'] ?? true) !== false) {
    fail('mid-apply second-file failure must not report success');
}
if (empty($mid['backup_metadata'])) {
    fail('mid-apply must create backup metadata for restore');
}
if (file_get_contents($fileA) !== $origA) {
    fail('file A must be restored from manifest after mid-apply failure');
}
if (file_get_contents($fileB) !== $origB) {
    fail('file B must be restored from manifest after mid-apply failure');
}

chdir($prevCwd);

// Cleanup best-effort
function rrmdir(string $dir): void {
    if (!is_dir($dir)) {
        return;
    }
    foreach (scandir($dir) ?: [] as $item) {
        if ($item === '.' || $item === '..') {
            continue;
        }
        $path = $dir . DIRECTORY_SEPARATOR . $item;
        if (is_dir($path)) {
            rrmdir($path);
        } else {
            @unlink($path);
        }
    }
    @rmdir($dir);
}
rrmdir($tmp);

fwrite(STDOUT, "OK apply_pipeline_test.php\n");
exit(0);
